pipelineRunner.go

```diff
@@ -50,712 +49,527 @@ func NewPipelineRunner(envvarHelper EnvvarHelper, whenEvaluator WhenEvaluator, d
-	result.LogLines = make([]contracts.BuildLogLine, 0)
+	if stage.ContainerImage != "" {
+		stage.ContainerImage = os.Expand(stage.ContainerImage, pr.envvarHelper.getEstafetteEnv)
+		isPulledImage = pr.dockerRunner.IsImagePulled(stage.Name, stage.ContainerImage)
+		isTrustedImage = pr.dockerRunner.IsTrustedImage(stage.Name, stage.ContainerImage)
 
-	log.Info().Msgf("[%v] Starting stage '%v'", p.Name, p.Name)
+		if !isPulledImage || runtime.GOOS == "windows" {
 
-	if p.ContainerImage != "" {
-		result.IsDockerImagePulled = pr.dockerRunner.isDockerImagePulled(p.Name, p.ContainerImage)
-		result.IsTrustedImage = pr.dockerRunner.isTrustedImage(p.Name, p.ContainerImage)
-
-		if !result.IsDockerImagePulled || runtime.GOOS == "windows" {
-
-			// log tailing - start stage as pending
-			status := "PENDING"
-			pr.tailLogsChannel <- contracts.TailLogLine{
-				Step:         p.Name,
-				ParentStage:  parentStageName,
-				Type:         "stage",
-				Depth:        depth,
-				RunIndex:     runIndex,
-				Image:        getBuildLogStepDockerImage(result),
-				AutoInjected: &result.Stage.AutoInjected,
-				Status:       &status,
+			buildLogStepDockerImage = &contracts.BuildLogStepDockerImage{
+				Name:      getContainerImageName(stage.ContainerImage),
+				Tag:       getContainerImageTag(stage.ContainerImage),
+				IsTrusted: isTrustedImage,
+				IsPulled:  isPulledImage,
 			}
 
+			// start pulling stage
+			pr.sendStatusMessage(stage, parentStageName, depth, runIndex, buildLogStepDockerImage, nil, contracts.StatusPending)
+
 			// pull docker image
 			dockerPullStart := time.Now()
-			result.DockerPullError = pr.dockerRunner.runDockerPull(ctx, p.Name, p.ContainerImage)
-			result.DockerPullDuration = time.Since(dockerPullStart)
-			if result.DockerPullError != nil {
-				return result, result.DockerPullError
-			}
+			err = pr.dockerRunner.PullImage(ctx, stage.Name, stage.ContainerImage)
+			imagePullDuration = time.Since(dockerPullStart)
 		}
 
 		// set docker image size
-		size, err := pr.dockerRunner.getDockerImageSize(p.ContainerImage)
-		if err != nil {
-			return result, err
+		if err == nil {
+			imageSize, err = pr.dockerRunner.GetImageSize(stage.ContainerImage)
+		}
+
+		if err == nil {
+			buildLogStepDockerImage = &contracts.BuildLogStepDockerImage{
+				Name:         getContainerImageName(stage.ContainerImage),
+				Tag:          getContainerImageTag(stage.ContainerImage),
+				IsTrusted:    isTrustedImage,
+				IsPulled:     isPulledImage,
+				ImageSize:    imageSize,
+				PullDuration: imagePullDuration,
+			}
 		}
-		result.DockerImageSize = size
 	}
 
-	// log tailing - start stage
-	status := "RUNNING"
-	pr.tailLogsChannel <- contracts.TailLogLine{
-		Step:         p.Name,
-		ParentStage:  parentStageName,
-		Type:         "stage",
-		Depth:        depth,
-		RunIndex:     runIndex,
-		Image:        getBuildLogStepDockerImage(result),
-		AutoInjected: &result.Stage.AutoInjected,
-		Status:       &status,
+	if err == nil {
```

Comment:
I think dealing with the err sooner will make the code cleaner

Reply:
It would but then we have to run the following in all those cases:
```
	// finalize stage
	finalStatus := contracts.StatusSucceeded
	if pr.canceled {
		log.Info().Msgf("%v Canceled pipeline '%v'", stagePlaceholder, stage.Name)
		finalStatus = contracts.StatusCanceled
	} else if err != nil {
		log.Warn().Err(err).Msgf("%v Pipeline '%v' container failed", stagePlaceholder, stage.Name)
		finalStatus = contracts.StatusFailed
	} else {
		log.Info().Msgf("%v Finished pipeline '%v' successfully", stagePlaceholder, stage.Name)
	}

	pr.sendStatusMessage(stage, parentStageName, depth, runIndex, nil, &runDuration, finalStatus)
```

Reply:
Replaced this with a defer statement, passing in the `err` as a pointer to be able to set status depending on it and return early everywhere.